File: monitoring/src/test/java/com/amazon/elasticsearch/resthandler/RestDeleteMonitorTests.java
/*
 * Copyright 2013-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 */
package com.amazon.elasticsearch.resthandler;

import com.amazon.elasticsearch.MonitoringPlugin;
import org.elasticsearch.action.admin.cluster.node.info.NodeInfo;
import org.elasticsearch.action.admin.cluster.node.info.NodesInfoResponse;
import org.elasticsearch.client.Client;
import org.elasticsearch.client.transport.TransportClient;
import org.elasticsearch.common.SuppressForbidden;
import org.elasticsearch.common.network.NetworkModule;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.common.transport.TransportAddress;
import org.elasticsearch.plugins.Plugin;
import org.elasticsearch.plugins.PluginInfo;
import org.elasticsearch.test.ESIntegTestCase;
import org.elasticsearch.test.junit.annotations.TestLogging;
import org.elasticsearch.transport.client.PreBuiltTransportClient;
import org.junit.Before;
import org.junit.Test;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Collection;

import static java.util.Arrays.asList;

@TestLogging("level:DEBUG")
@ESIntegTestCase.ClusterScope(scope=ESIntegTestCase.Scope.SUITE, numDataNodes=3)
public class RestDeleteMonitorTests extends ESIntegTestCase {

    @Before
    public void setUp() throws Exception {
        super.setUp();
    }

    @Override
    protected Collection<Class<? extends Plugin>> nodePlugins() {
        return asList(MonitoringPlugin.class);
    }

    public void testPluginIsLoaded() {
        NodesInfoResponse response = client().admin().cluster().prepareNodesInfo().setPlugins(true).get();
        for (NodeInfo nodeInfo : response.getNodes()) {
            boolean pluginFound = false;
            for (PluginInfo pluginInfo : nodeInfo.getPlugins().getPluginInfos()) {
                if (pluginInfo.getName().equals("aes-monitoring")) {
                    pluginFound = true;
                    break;
                }
            }
            assertTrue(pluginFound);
        }
        ensureYellow();
    }

    public void testGetAlertEmpty() {
        Client client = client();
//        Settings settings = Settings.builder().put("client.transport.ignore_cluster_name",true).build();
//        Client client = client();
//        Response response = getRestClient().performRequest("GET",
//                "/_awses/policies",
//                Collections.emptyMap());
//        GetRequest request = new GetRequest();
//        client().get(request);
//        getRestClient();
//        this.restClient = getRestClient();
//        getRestClient();
//        MonitorManager policyManager = new MonitorManager(getRestClient(), logger);
//        this.restClient = RestClient.builder(new HttpHost("localhost", 55716, "http")).build();
//
//        Response response;
//        response = restClient.performRequest(RestRequest.Method.GET.toString(), POLICY_INDEX_PATH+ "_search");
    }
}

File: monitoring/src/test/java/com/amazon/elasticsearch/resthandler/RestDeleteMonitorTests.kt
/*
 * Copyright 2013-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 */
package com.amazon.elasticsearch.resthandler

import com.amazon.elasticsearch.monitor.Monitor
import org.apache.http.entity.ContentType
import org.apache.http.entity.StringEntity
import org.elasticsearch.common.xcontent.NamedXContentRegistry
import org.elasticsearch.common.xcontent.XContentFactory
import org.elasticsearch.common.xcontent.XContentType
import org.elasticsearch.rest.RestStatus
import org.elasticsearch.test.ESTestCase
import org.elasticsearch.test.junit.annotations.TestLogging
import org.elasticsearch.test.rest.ESRestTestCase

@TestLogging("level:DEBUG")
class RestDeleteMonitorTests : ESRestTestCase() {

    @Throws(Exception::class)
    fun testPluginIsLoaded() {
        val response = entityAsMap(ESRestTestCase.client().performRequest("GET", "_nodes/plugins"))
        val nodesInfo = response["nodes"] as Map<String, Map<String, Any>>
        for (nodeInfo in nodesInfo.values) {
            val plugins = nodeInfo["plugins"] as List<Map<String, Any>>
            for (plugin in plugins) {
                if (plugin["name"] == "aes-monitoring") {
                    return
                }
            }
        }
        fail("Plugin not installed")
    }

    @Throws(Exception::class)
    fun testDeleteMonitor() {
        val monitor = createRandomMonitor()

        val deleteResponse = client().performRequest("DELETE", "_awses/monitors/" + monitor.id)
        assertEquals("Delete did not return correct response",
                RestStatus.OK, RestStatus.fromCode(deleteResponse.statusLine.statusCode))

        val getResponse = client().performRequest("HEAD", "_awses/monitors/" + monitor.id)
        assertEquals("Deleted monitor is still present",
                RestStatus.NOT_FOUND, RestStatus.fromCode(getResponse.statusLine.statusCode))
    }

    private fun createRandomMonitor() : Monitor {
        val monitor = randomMonitor()
        val monitorSource = monitorSource(monitor)
        val response = client().performRequest("POST", "_awses/monitors",
                emptyMap(), StringEntity(monitorSource, ContentType.APPLICATION_JSON))
        assertTrue("Unable to create a new monitor", response.statusLine.statusCode == RestStatus.CREATED.status)

        val monitorId = XContentType.JSON.xContent()
                .createParser(NamedXContentRegistry.EMPTY, response.entity.content)
                .map()
                .get("_id") as String
        return monitor.copy(id = monitorId)
    }

    private fun monitorSource(monitor: Monitor): String {
        val builder = XContentFactory.jsonBuilder()
        monitor.toXContent(builder)
        return builder.string()
    }

    private fun randomMonitor(): Monitor {
        return Monitor(id = Monitor.NO_ID,
                name = randomAlphaOfLength(10),
                enabled = ESTestCase.randomBoolean(),
                search = "not implemented",
                schedule = "not implemented",
                actions = listOf());
    }
}
